### DISCUSSION TITLE
LNK2005 Error in Microsoft Visual Studio 2022 using CGAL_LINKED_WITH_3MF

### DISCUSSION THREAD

**Discussion by henriqueramalho1** [CGAL/cgal#7462] [Q&A]
Hello, I'm having some trouble to compile my code to open 3mf models using CGAL. I've already set up Lib3MF and all the includes, but when I try to compile, I get LNK2005 error. The functions that the error reports are in "write_3mf.h" and "read_3mf.h". As they are defined in headers, I thought that this could be the cause, but looking in other functions in CGAL, there are more functions defined in headers and they are not causing LNK2005.

I tried to add "inline" in all these functions, then it worked, but I want to know how to fix this problem without editing library headers.

My code is like that:

(MeshBuilder.h)

```
#pragma once
#include "Mesh_DCEL.h"
#include <CGAL/Surface_mesh/Surface_mesh.h>
#include <CGAL/Exact_predicates_inexact_constructions_kernel.h>

typedef CGAL::Exact_predicates_inexact_constructions_kernel  Kernel;
typedef CGAL::Surface_mesh<Kernel::Point_3>                      SurfaceMesh;

class MeshBuilder
{
private:
	SurfaceMesh mesh;
	Mesh_DCEL* dcel;

public:
	MeshBuilder(Mesh_DCEL* m);
	~MeshBuilder();
	bool build(std::string path);
	void clone();
};
```

(main.cpp)

```
#include <iostream>
#include "Mesh_DCEL.h"
#include "MeshParserStrategy.h"
#include "IncrementalSlicer.h"
#include "SolidSlice.h"
#include "MeshBuilder.h"
#include "Translate3D.h"
#include <stdlib.h>
#include <string.h>

int main(int argc, char** argv)
{
    ...
}
```

![erro1](https://github.com/CGAL/cgal/assets/82188037/c7467170-8b6f-4099-b6d0-8ab4731b79c7)

**Comment by afabri** [comment]
Note that function [`CGAL::IO::read_3MF()`](https://doc.cgal.org/latest/Surface_mesh/group__PkgSurfaceMeshIOFunc3MF.html#ga1131e46ab505eea6b80a4d952070074d)  is for an older version of lib3mf.

Before upgrading, can you tell me if it is worth it?  That is, how successful is this data format?

**Reply by henriqueramalho1** [reply]
Hello, Afabri. I'm already using Lib3mf 1.8.1, that is inferior to 2.0. The data format 3MF is good for manufacturing planning process, because 3D models triangles and vertices are described unically. Furthermore, it can store colors and it makes it easy to build a topological representation of the model.

**Comment by afabri** [comment]
I guess there is too much code that has nothing to do with CGAL.    Did you git clone CGAL?    In  Surface/mesh/test/Surface_mesh/ you have one testcase that uses write_3MF and that links against the library.  Maybe try to compile that first.

**Reply by henriqueramalho1** [reply]
Following your suggestions, I compiled the test and it worked well! 

This test runs nicely

```
#include <CGAL/Installation/internal/disable_deprecation_warnings_and_errors.h>
#include <fstream>
#include <iostream>
#include <CGAL/Surface_mesh.h>
#include <CGAL/Simple_cartesian.h>
#include <CGAL/boost/graph/generators.h>
#include <CGAL/Surface_mesh/IO/3MF.h>
#include <CGAL/Surface_mesh/IO/OFF.h>
#include <CGAL/Surface_mesh/IO/PLY.h>


typedef CGAL::Simple_cartesian<double>         Kernel;
typedef Kernel::Point_3                        Point_3;
typedef CGAL::Surface_mesh<Point_3>            SM;

int main()
{
    SM sm_in;
    std::vector<SM> output_3mf;
    bool ok = CGAL::read_3mf("test.3mf", output_3mf);
    assert(ok);
    assert(output_3mf.size() == 2);
    sm_in.clear();
}
```

Then I did a test. I moved the test code to a simple class

MeshBuilder.h
```
#pragma once

#include <CGAL/Installation/internal/disable_deprecation_warnings_and_errors.h>
#include <CGAL/Surface_mesh.h>
#include <CGAL/Simple_cartesian.h>
#include <CGAL/boost/graph/generators.h>
#include <CGAL/Surface_mesh/IO/3MF.h>
#include <CGAL/Surface_mesh/IO/OFF.h>
#include <CGAL/Surface_mesh/IO/PLY.h>

typedef CGAL::Simple_cartesian<double>         Kernel;
typedef CGAL::Surface_mesh<Kernel::Point_3>                      SurfaceMesh;

class MeshBuilder
{
private:
	SurfaceMesh mesh;

public:
	MeshBuilder();
	~MeshBuilder();
	bool build(std::string path);
};
```
MeshBuilder .cpp
```
#include "MeshBuilder.h"

MeshBuilder::MeshBuilder():
	mesh()
{
	
}

MeshBuilder::~MeshBuilder()
{
}

bool MeshBuilder::build(std::string path)
{
	std::vector<SurfaceMesh> output;
	if (CGAL::IO::read_3MF(path, output))
	{
		mesh = output[0];

		std::cout << "Opened 3MF successfully" << std::endl;
		return true;
	}
	else
		return false;
}
```

Then I included this in main.cpp

```
#include "MeshBuilder.h"

int main()
{
    MeshBuilder builder;
    builder.build("test.3mf");
}
```

Then I got the same **LNK2005** error, just moving the code to a class. 

I did another observation, the functions that are responsible for LNK2005 error are not "template", for example this three functions in "write_3mf.h"
![threeFunc](https://github.com/CGAL/cgal/assets/82188037/a278790e-4cd0-4dbf-95c0-9ce8cf3a2544)

Other functions like this, that has "template" are free of LNK errors.
![oneFunc](https://github.com/CGAL/cgal/assets/82188037/b73491f3-2ba8-45fb-8814-69bf140e982f)

But anyway, I add "inline" in those fucntions that are causing LNK2005 and it worked well!
![inline](https://github.com/CGAL/cgal/assets/82188037/74b18037-27ea-4fa7-ba91-83a4aca677ac)

**Reply by lrineau** [reply]
Hi Henrique, you might learn more about the reasonning of that error lnk2005 at at  https://learn.microsoft.com/en-us/cpp/error-messages/tool-errors/linker-tools-error-lnk2005?view=msvc-170.
